mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,621 @@
+            "unsampled": kwargs["unsampled"],
+            "preds_single": preds_single,
+            "preds_multi": preds_multi,
+            "do_rank": True,
+            "return_dict": True,
+        }
+
+        strategies = ["lcs", "ms", "es"]
+        strategies = strategies + ["ds", "kld"] if preds_multi else strategies
+
+        for strategy in strategies:
+            ranked_inds = sampling_methods[strategy].sample(**heuristic_params)[
+                "ranked_indices"
+            ]
+            for rank, qid in enumerate(ranked_inds):
+                qid2rank[qid] += rank
+            num_strategies_used += 1
+
+        qid2rank = [(k, v / num_strategies_used) for k, v in qid2rank.items()]
+        all_qids, all_ranks = list(zip(*qid2rank))
+
+        reordered = custom_reordering(
+            all_ranks, do_rank=False, sampling_size=sampling_size
+        )
+        ranked_indices = [all_qids[i] for i in reordered]
+        sampled_indices, unsampled_indices = (
+            ranked_indices[:sampling_size],
+            ranked_indices[sampling_size:],
+        )
+        return sampled_indices, unsampled_indices, ranked_indices
+
+
+class KLDivergenceSampling(Heuristic):
+    """Selection strategy that calculates the KL divergence between the posterior
+    distribution of multiple classifiers."""
+
+    def _extractor(
+        self,
+        sampling_size: int,
+        preds_multi: List[List[List[float]]],
+        do_rank: bool = True,
+        domain_indices: Dict = None,
+        **kwargs,
+    ):
+        """
+        Args:
+            preds_multi (List[List[float]]): Probability scores for each data point for each
+                class from multiple classifiers.
+            do_rank: if True, returns a ranked list of the indices for confidence-sorting of data
+                samples.
+            domain_indices (Dict[str, tuple(int, int)]): A mapping between domains (str) to the
+                corresponding indices in the probability vector. Used for intent-level KLD.
+        """
+        sampling_size = Heuristic.validate_sampling_size(sampling_size)
+
+        # Calculate average prediction values.
+        avg_preds = None
+        for pred in preds_multi:
+            if avg_preds is None:
+                avg_preds = np.array(pred, dtype=np.float64)
+            else:
+                avg_preds += np.array(pred, dtype=np.float64)
+        avg_preds /= len(preds_multi)
+        # Estimate divergence.
+        divergences = None
+        if not domain_indices:
+            divergences = KLDivergenceSampling.get_divergence_all_domains(
+                preds_multi, avg_preds
+            )
+        else:
+            divergences = KLDivergenceSampling.get_divergences_within_domain(
+                preds_multi, avg_preds, domain_indices
+            )
+        # after transpose: row -> data point, column -> model
+        divergences = np.array(divergences).T
+        confidences = -np.max(divergences, axis=1)
```

Comment:
Could you add a comment on why the negative is taken here. I am guessing it is something to do with the entropy function

Reply:
The negative was taken here to change the order of the sorting. However, this is a double negative and has been fixed to reorder the ordered indices using [::-1] instead.

```
        entropy_per_element = scipy_entropy(
            np.array(confidences_2d), axis=1, base=ENTROPY_LOG_BASE
        )
        high_to_low_entropy = np.argsort(entropy_per_element)[::-1]
        return list(high_to_low_entropy)
```